### DISCUSSION TITLE
[BP#15] Choisir les technologies les plus adaptées

### DISCUSSION THREAD

**Discussion by devcrafting** [cnumr/best-practices#199] [Bonnes pratiques] [answered]
Comme pointé par @florinesueur à propos de la BP 11, je pense que celle-ci est plus claire, bien qu'il y ait quelques références un peu datées ou du wording trop restrictif, et un exemple trop simple ?

Peut-être qu'on peut reprendre des exemples ?

La mesure me semble compliquée, vu que l'appréciation de "inopportune" est très subjective. Peut-être parler de taux d'utilisation des différentes fonctionnalités ? Cela propose au moins un cadre d'analyse pour se demander quelle est la bon niveau d'abstraction

**Comment by devcrafting** [comment]
A noter le lien avec https://github.com/cnumr/best-practices/discussions/206 et https://github.com/cnumr/best-practices/discussions/77

Je me demande si on n'a pas intérêt à expliciter chacun des 5 cas dans 5 BP pour mettre l'emphase sur ces 5 choix et pouvoir expliciter un peu plus les critères de choix ?

**Comment by devcrafting** [comment]
Mentionné ici également https://github.com/cnumr/best-practices/discussions/122
Ne serait-ce pas un autre axe que les 5 axes plutôt orientés back...
Avec la proposition, on serait plutôt sur du front : plus ou moins intéractif/adaptatif au terminal ?

**Reply by florinesueur** [reply]
Oui en effet, je me pose aussi la question pour la PWA.

**Comment by florinesueur** [comment]
Peut-être faudrait-il un exemple (des cas d'usage) pour chacune des technologies  ?

À vrai, dire le principe de validation actuel, je ne le comprends pas vraiment.
Est-ce qu'un taux d'utilisation pourrait-être intéressant ? Par exemple, j'utilise à 30% les possibilités natives de wordpress qui est à mon sens une justification possible pour du jamstack ou du statique sur mesure. 

Ou sinon que les technologies utilisées sont indispensables dans le sens où il serait impossible de le faire avec une technologie à moindre impact dans le classement et en fonction de son savoir-faire (sinon les cms passent toujours à la trappe).

**Reply by ldesgrange** [reply]
Un exemple par type de techno me semble être une bonne idée. En tout cas l'exemple actuel dan la BP n'est pas assez explicite.

Pour la règle de validation, un taux d'utilisation des possibilité d'une technologie ne me semble pas clair non plus (et dans ce cas le taux change probablement d'une version à l'autre de la techno ?).
Par contre, effectivement, remplacer "inopportunes" par "non-indispensables" ou "irremplaçable par une techno à moindre impact" me paraît déjà plus claire.

**Reply by devcrafting** [reply]
Et si on découpe en BP distincte du genre "Choisir un CMS quand c'est adapté", "Choisir une architecture statique quand c'est adapté"...?
Ca permettrait d'expliciter des critères plus clairs pour chacun des 5 types identifiés dans la BP

**Reply by ACTLEM** [reply]
L'un de vous (ou plusieurs)serait chaud pour proposer dans cette discussion une liste de BP: titre + argumentaire succint pour remplacer, détailler cette BP?

**Reply by ACTLEM** [reply]
Ou sinon tout lister dans une BP comme l'a fait plus bas Florine, mais j'ai peur que la BP devienne trop "fat".

**Reply by florinesueur** [reply]
Oui c'est trop "fat" surtout qu'une page de livre n'est pas scrollable.

**Reply by devcrafting** [reply]
J'aime bien la liste de @florinesueur peut-être à quelques précisions près. On pourrait faire un point 15-30min semaine prochaine ?

**Reply by ACTLEM** [reply]
Salut,

A défaut de faire un point, @florinesueur @devcrafting, c'est possible de lister les BP qui remplaceront cette bonne pratique ainsi que la BP 11 (https://github.com/cnumr/best-practices/discussions/77) qui est CMS vs sur Mesure?
De ce que je vois pour faire avancer le débat (version "martyre" ^^):
- "Favoriser un site web statique" (-> BP13): Si le site ne contient pas de données dynamiques, peu voire pas de mises à jour ("site plaquette")
- "Privilégier l'utilisation d'un CMS pour des sites à contenu": Site avec du contenu à manager, éviter de réinventer la roue. Parler de Headless CMS (ex: Strapi de plus en plus utilisé)
- "Privilégier des PWA pour des sites couplés à une application mobile": Voir discussion https://github.com/cnumr/best-practices/discussions/122
- "Privilégier un site dynamique sur mesure pour des services complexes" (preneur de meilleur titre): Site avec problématiques complexes, genre e-commerce, location en ligne etc...
- "Privilégier des frameworks pour des sites exigeants": Site avec des problématiques fortes de performances et des formulaires complexes et de la manipulation de données, éviter de réinventer la roue (rejoint un peu la https://github.com/cnumr/best-practices/discussions/208)

**Reply by florinesueur** [reply]
Je reprends ce que tu as marqué @ACTLEM dans un tableau en complétant un peu pour y voir un peu plus clair.

| En remplacement de la BP  | Titre         | Description |Exemples d'utilisation | Exemples de technologies |
| :--------------- |:---------------|:-----|:-----| :-----|
| [BP#13](https://github.com/cnumr/best-practices/blob/main/chapters/BP_013_fr.md) Discussion #206  |   Favoriser un site web statique | Site non dynamique avec pas ou peu de mise à jour  |  Site vitrine, plaquette, Landing page, Blog,... | Statique : Html, Css, JS / Statique généré : Hugo, Jekyll, Gatsby, Next, Nuxt,... |
| [BP#11](https://github.com/cnumr/best-practices/blob/main/chapters/BP_011_fr.md) Discussion #77   | Privilégier l'utilisation d'un CMS pour des sites à contenu  |      Site avec du contenu à manager     |  Blog, E-commerce,... | CMS : Wordpress, Drupal, joomla, Shopify ... / Headless CMS : Strapi, Contenful, NetlifyCMS, ... |
| New Discussion #122  | Privilégier des PWA pour des sites couplés à une application mobile       |  Au lieu d'un duplicat de site avec une application native |  E-commerce, Saas, ... | Angular, React, Vue, Svelte, Ionic, ... |
| New  | Privilégier un site dynamique sur mesure pour des services complexes    | Pour des besoins spécifiques nécessitant un contrôle total dépassant les contraintes des frameworks ? | unité fonctionnelle métier spécifique ? | PHP, J2EE, .NET, Rust, ... |
| New  | Privilégier des frameworks pour des sites exigeants  | Site avec des problématiques fortes de performances et des formulaires complexes et de la manipulation de données | SaaS (suivi de la bourse) ? | React, Vue, Svelte, Rocket, ... |

Je ne comprends pas bien dans quel cas, il faudrait privilégier un site dynamique sur-mesure vs framework.

**Reply by tbroyer** [reply]
J'imagine que ça dépend de ce qu'on met derrière "framework".

Si on y met Spring, Symfony, Rails, Django etc. (ou JavaEE), alors je ne souscris pas du tout à ce qui est dit (carrément l'inverse). De tels frameworks viennent avec plein de points d'extension (complexité accrue) et/ou de fonctionnalités dont on n'aura pas besoin, on ne peut les privilégier que pour la productivité des développeurs ou l'harmonisation de la stack technique entre les projets (et on obtient des "développeurs <framework X>" au lieu de "développeurs <langage Y>"; comme on a pu avoir des "développeurs jQuery" à une époque; mais c'est un autre sujet).

Mais surtout les deux dernières lignes du tableau sont pour moi complémentaires, pas opposées : utiliser un framework **c'est** faire du développement sur-mesure, vs utiliser un CMS ou, pour un site e-commerce, un Shopify ou Magento.
(ceci étant, telle que décrite par Thomas, je ne comprend vraiment pas cette dernière BP sur les frameworks)

**Reply by ACTLEM** [reply]
Merci @florinesueur pour la mise en format tableau, c'est plus clair. 
Je pensais plus à React, Svelte quand je parlais de frameworks, j'étais plus côté UI (pour le constater et le subir régulièrement, je rejoins à 100% @tbroyer sur sa remarque concernant Spring, Symfony etc....) 
Effectivement, en ayant le format tableau, ça me saute aussi aux yeux que les deux derniers points se recoupent (sites complexes, exigeants en terme de performance par exemple...). Pour cela, je pense beaucoup à des expériences e-commerce avec fort traffic et un peu spécifiques.

**Reply by florinesueur** [reply]
Si je vous comprends bien les deux dernières lignes pourraient être une unique BP ?

**Comment by fzaninotto** [comment]
Je vois de nombreux contre-exemples à cette BP. 

- Un framework dispose souvent d'optimisations qui améliorent sa performance (request pooling, cache, etc). 
- Un gros CMS a toujours une couche de prégénération (ou un gros cache), et l'affichage des pages revient souvent à la recomposition d'ESIs, qui est bien plus rapide que ce qu'on pourrait faire en PHP par exemple.
- Un site développé sans framework n'a aucune garantie d'être économe. 
- Il n'y a pas de différence perceptible de consommatin de ressources entre un site développé à la main (ça existe encore?) et un site généré statiquement avec n outil comme Jekyll.  

Pour donner un avis objectif, il faudrait pouvoir comparer l'impact d'un même site développé avec deux architectures différentes. Or il existe très très peu d'études en ce sens (même RealWorld ne compare que des technos différentes sur une architecture commune, la SPA)., et aucune à ma connaissance qui donne un avantage clair à l'une ou l'autre architecture. 

Bref, cette BP très vague et souvent fausse n'apporte rien et décrédibilise, à mon avis, les autres. 

Je propose une reformulation en "Générer des pages statiques à la publication time plutôt que les calculer à la demande". Là, on sait comparer, et je pense qu'il n'y a pas de doute sur la différence.

**Reply by devcrafting** [reply]
A mon sens, c'est utile pour ne pas forcément foncer sur WordPress pour faire un site statique ou quasi-statique (j'ai poussé cette approche avec ma femme pour son site, en Hugo.io - equiv Jekyll - où elle veut pouvoir faire quelques mises à jour : elle publie du contenu sur GitLab, simplifier avec Markdown, c'est intégré/publié en déploiement continu).
Sur des startups, un événement, j'ai déjà fait des landings en full HTML (mais ok que la diff avec un générateur type Jekyll il n'y a pas de grande différence : ça ne joue que sur les ressources nécessaires à la génération/publication, soit assez limité). 

Ce qui me paraît vraiment important c'est d'avoir des questions à se poser pour utiliser une solution technique de manière éclairée et non pas par mode/par défaut. Et ma proposition est de peut-être découper en plusieurs BP par typologie d'architecture (5 identifiées dans cette BP)

**Reply by florinesueur** [reply]
Oui, je suis d'accord avec @devcrafting . Du coup, nous ne sommes plus dans un classement, mais dans un listing de possibilités avec des questionnements et indications pour chaque typologie d'architecture ?

- site statique (HTML/CSS/JS) ;
- site dynamique (PHP, J2EE, .NET, Rust, etc.) ;

- site statique généré (Jekyll, Hugo, Gatsby, etc.) ;
- site hybride (statique et dynamique) (Next, Nuxt, etc.) ;
- site dynamique avec un framework (Symfony, React, Vue, Svelte, Rocket, etc.) ;
- Progressive Web App (React, Vue, Svelte, Ionic, etc.);
- Application Desktop (electron, etc.);

- site dynamique avec un CMS (Wordpress, Drupal, Joomla!, Jahia, etc.);
- site hybride (Gatsby, Next, etc.) avec un Headless CMS (Wordpress, Contenful, GraphCMS etc.).

Voyez-vous d'autres typologies ou des typologies à enlever (discussion sur la PWA la semaine prochaine) ou d'autres exemples ?
J'enlèverai "développer" et aussi "sur-mesure" car c'est le cas pour toutes les typologies à part le CMS.

**Reply by tbroyer** [reply]
"Progressive Web App" → "Single Page App"

On peut très bien faire une PWA qui ne soit pas une SPA.
* https://hpbn.co/ tombe dans la catégorie "PWA" alors que c'est un site statique.
* on peut faire du templating dans le service worker et avoir une "vraie MPA" concernant la navigation

Concernant les SPA et les MPA, il y a aussi tout le spectre entre les 2 (https://www.youtube.com/watch?v=860d8usGC0o)

**Reply by rlemaire** [reply]
La bonne technologie dépend du contexte et des contraintes du projets. Une SPA très lourde peut être adaptée dans un cas particulier par exemple.
Comme le montre les discussions il n'y a pas de réponses absolue.
C'est intéressant je pense d'avoir une pratique qui dit « est-ce que vous avez choisi la meilleure option technique ? » sans forcement donner une règle absolue, mais avec des exemples d'alternatives à considérer.

**Comment by ACTLEM** [comment, marked as answer]
@tbroyer J'ai créé l'issue pour modification (ou suppression et ajout) pour parler d'une alternative adaptée quand pas statique et sur-mesure (tu avais parlé des SPA): https://github.com/cnumr/best-practices/issues/329
@devcrafting @rlemaire @florinesueur si vous avez une idée de BP sur ce thème (j'avais pensé un moment aux CMS Headless mais pas trouvé de formulation adéquate), n'hésitez pas à proposer une PR en référence à cette discussion ou à l'issue.
